### PR TITLE
Added curl-patch tool

### PR DESCRIPTION
**Pull request by devgianlu**
This tool allows to log all the requests made with `libcurl` to the console.

### CODE REVIEW

File: curl-patch/Makefile [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+DYLIB=curl_patch.dylib
+INPUT=curl_patch.c
+
+curl_patch:
+	$(CC) $(CFLAGS) -o $(DYLIB) $(INPUT)
+
+.PHONY: clean
+
+clean:
+	rm -f $(DYLIB) $(INPUT)
```
Comment (kingosticks) [inline curl-patch/Makefile]: You really want `make clean` to remove curl_patch.c?!

File: curl-patch/Makefile [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+	CFLAGS += -fPIC -shared
+endif
+ifeq ($(UNAME_S),Darwin)
+	CFLAGS += -dynamiclib
+endif
+
+DYLIB=curl_patch.dylib
+INPUT=curl_patch.c
+
+curl_patch:
```
Comment (kingosticks) [inline curl-patch/Makefile]: `curl_patch: curl_patch.c` ?

**Comment by kingosticks** [comment]
Does the libcurl stuff really not end up in the wireshark dump?

**Comment by devgianlu** [comment]
It's HTTPS, so it is not very clear what's happening.

File: README.md [hunk 3]
```diff
@@ -19,6 +19,15 @@ This will produce a `dump.pcap` in the current directory, which can be analyzed
 ## spotify-dissect
 The `spotify-dissect` tool is a set of Wireshark dissectors used to analyze packet dumps.
 
+## spotify-curl-patch
+The `spotify-curl-patch` tool sets `CURLOPT_VERBOSE` to `1` logging to the console every request made with `libcurl`.
```
Comment (kingosticks) [inline README.md:32]: You should explain that this lets you (easily) see the unencrypted requests.

File: README.md [hunk 4]
```diff
@@ -19,6 +19,15 @@ This will produce a `dump.pcap` in the current directory, which can be analyzed
 ## spotify-dissect
 The `spotify-dissect` tool is a set of Wireshark dissectors used to analyze packet dumps.
 
+## spotify-curl-patch
```
Comment (kingosticks) [inline README.md:31]: Personally I'd have called this something like 'spotify-dump-http'.

**Comment by plietar** [comment]
Could this be merged with the existing spotify-dump instead?

**Comment by devgianlu** [comment]
~I'd need to create a curl to pcap script.~ I can use `CURLOPT_DEBUGFUNCTION`, I'll give it a try.

File: dump-http/patch.sh [hunk 5]
```diff
@@ -0,0 +1,4 @@
+#!/usr/bin/env bash
```
Comment (kingosticks) [inline dump-http/patch.sh:1]: dump-http/patch.sh needs to be executable. It's not.

**Comment by devgianlu** [comment]
Writing a `pcap` file is a bit too complicated for my C skills. I'll leave this to someone else.

**Comment by sashahilton00** [comment]
Can we not write this the the existing pcap file that the mercury stuff gets dumped to? I haven't looked at pcap format, but it was my understanding that it can store multiple dumps of different protocols?

**Comment by devgianlu** [comment]
I'm pretty sure that can be done. @kingosticks may be able to do that.

**Comment by kingosticks** [comment]
Ideally wouldn't we want both this and the protobuf dumping for a given run? I guess they could each dump to different pcap files and then wireshark can load them both and presumably lineup all the timestamps properly. That'd be cool.

**Comment by sashahilton00** [comment]
I think having it all in one file w/ timestamps would be optimal. 2 seperate files sounds unnecessary. I've been looking into ways to get a realtime capture (on a side note), and will try and add support for FIFO pipes when I get some spare time. Reading over https://wiki.wireshark.org/CaptureSetup/Pipes, it doesn't look like we need to change much to add support to `dump.c`. What I'd ideally like to do is merge dump and dissect into one, so I can just run `./proxy /Applications/Spotify.app/Contents/MacOS/Spotify` and have Wireshark load up and give me a realtime stream of packets, though I appreciate that most people probably want `dump` and `dissect` kept seperate.

**Comment by kingosticks** [comment]
I agree but I'd imagine the http stuff will be running in a different thread to the protobuf stuff so we'll have some synchronisation issues when writing to the same file. Maybe they are easy to solve but two files eliminates it entirely.

I've been using it lately where I start the dissector while I'm still capturing and it works fine. Except you do have to click the "reload pcap" button in wireshark everytime you want to see the latest dumped stuff.

**Comment by sashahilton00** [comment]
> I've been using it lately where I start the dissector while I'm still capturing and it works fine. Except you do have to click the "reload pcap" button in wireshark everytime you want to see the latest dumped stuff.

Yeah that does work, but feels a bit clunky. WRT the threads issue, I suspect you are right, though one would have to investigate further to see if that was a significant issue or whether we could resolve it with timestamped events.

**Comment by k3d3** [comment]
Spotify appears to support SSLKEYLOGFILE (I'm not sure if this is from curl or electron, but a lot of traffic appears to be decrypted), so I think that may be the simplest way to get at the traffic.

**Comment by devgianlu** [comment]
@k3d3 I've just tried this, but I still see a lot of encrypted traffic.
image

**Comment by devgianlu** [comment]
I've actually managed to decrypt all the traffic, I wasn't able to find it because Spotify uses HTTP2. @k3d3 Thank you a lot!

**Comment by sashahilton00** [comment]
I'm going to merge this as is. There were a couple of other enhancements raised in this issue wrt the pcap file stuff and a possible live intercept, but those can be handled separately.

**Pull request merged**